service/src/main/kotlin/fi/espoo/evaka/reports/MealReportController.kt

```diff
@@ -0,0 +1,232 @@
+        PREPARATORY_DAYCARE,
+        PREPARATORY_DAYCARE_ONLY
+    )
+
+enum class MealType {
+    BREAKFAST,
+    LUNCH,
+    LUNCH_PRESCHOOL,
+    SNACK,
+    SUPPER,
+    EVENING_SNACK
+}
+
+fun mealtypeToMealIdTranslator(mealType: MealType): Int =
+    when (mealType) {
+        MealType.BREAKFAST -> 162
+        MealType.LUNCH -> 175
+        MealType.LUNCH_PRESCHOOL -> 22
+        MealType.SNACK -> 152
+        MealType.SUPPER -> 27
+        MealType.EVENING_SNACK -> 30
+    }
+
+val mealIdsSpecialDiet =
+    mapOf(
+        MealType.BREAKFAST to 143,
+        MealType.LUNCH to 145,
+        MealType.LUNCH_PRESCHOOL to 24,
+        MealType.SNACK to 160,
+        MealType.SUPPER to 28,
+        MealType.EVENING_SNACK to 31
+    )
+
+data class MealReportRow(
+    val mealType: MealType,
+    val mealId: Int,
+    val mealCount: Int,
+    val dietId: Int? = null,
+    val additionalInfo: String? = null
+)
+
+data class MealReportData(
+    val date: LocalDate,
+    val unitName: String,
+    val meals: List<MealReportRow>
+)
+
+data class MealInfo(
+    val mealType: MealType
+    // TODO to support special diets add extra info: additionalInfo (name), dietId
+)
+
+private fun childMeals(
+    fixedScheduleRange: TimeRange?,
+    reservations: List<ReservationResponse>,
+    absent: Boolean,
+    daycare: Daycare,
+    usePreschoolMealTypes: Boolean
+): Sequence<MealInfo> {
+    // if absent -> no meals
+    if (absent) {
+        return emptySequence<MealInfo>()
+    }
+    // list of time ranges when child will be present according to fixed schedule or reservation
+    // times
+    val presentTimeRanges =
+        if (fixedScheduleRange != null) listOf(fixedScheduleRange)
+        else reservations.filterIsInstance<ReservationResponse.Times>().map { it.range }
+    // if we don't have data about when child will be present, default to breakfast + lunch + snack
+    if (presentTimeRanges.isEmpty()) {
+        return sequenceOf(
+            MealInfo(MealType.BREAKFAST),
+            MealInfo(MealType.LUNCH),
+            MealInfo(MealType.SNACK)
+        )
+    }
+    // otherwise check unit meal times against the present time ranges
+    val meals = mutableListOf<MealInfo>()
+
+    fun addMealIfPresent(mealTime: TimeRange?, mealType: MealType) {
+        if (mealTime != null && presentTimeRanges.any { it.overlaps(mealTime) }) {
+            meals.add(MealInfo(mealType))
+        }
+    }
+
+    addMealIfPresent(daycare.mealtimeBreakfast, MealType.BREAKFAST)
+    addMealIfPresent(
+        daycare.mealtimeLunch,
+        if (usePreschoolMealTypes) MealType.LUNCH_PRESCHOOL else MealType.LUNCH
+    )
+    addMealIfPresent(daycare.mealtimeSnack, MealType.SNACK)
+    addMealIfPresent(daycare.mealtimeSupper, MealType.SUPPER)
+    addMealIfPresent(daycare.mealtimeEveningSnack, MealType.EVENING_SNACK)
+
+    return meals.asSequence()
+}
+
+private fun getMealReport(db: Database.Read, date: LocalDate, unitId: DaycareId): MealReportData {
```

Comment:
Transaktio-objektin muuttujanimi on yleisesti `tx`:

```suggestion
private fun getMealReport(tx: Database.Read, date: LocalDate, unitId: DaycareId): MealReportData {
```